### ISSUE TITLE
ID must be unique for UmlClassDiagram

### ISSUE THREAD

**Issue by clue** [clue/graph-uml#2]
> Migrated from clue/graph#18, reported by @clemens-tolboom

I somehow needed to do

``` php
$ git diff
diff --git a/lib/Fhaculty/Graph/Loader/UmlClassDiagram.php b/lib/Fhaculty/Graph/Loader/UmlClassDiagram.php
index 63e5035..2d15650 100644
--- a/lib/Fhaculty/Graph/Loader/UmlClassDiagram.php
+++ b/lib/Fhaculty/Graph/Loader/UmlClassDiagram.php
@@ -65,7 +65,7 @@ class UmlClassDiagram extends Base
         } else {
             $reflection = new ReflectionClass($class);
         }
-        $vertex = $this->graph->createVertex($class);
+        $vertex = $this->graph->createVertex($class, TRUE);

         $parent = $reflection->getParentClass();
         if ($parent) {
```

I feeded it with get_declared_classes();

``` php
  $loader = new UmlClassDiagram();
  try {
    foreach ($classes as $class) {
      $loader->createVertexClass($class);
    }
  }
  catch (Exception $exc) {
    return $exc->getMessage();
  }
```

My test environment is a little tricky so I was wondering how we test this easier.

BTW awesome UML generator.
- http://build2be.com/content/views-core-awesome-uml-route
- http://build2be.com/content/graphviz-blockstorageunittest

**Comment by clemens-tolboom** [comment]
I'm running into this again. Only something different happens as I get duplicate classes for classes having children.

Swapping the order of

```
$builder->createVertexClass('Fhaculty\Graph\Edge\Base');
$builder->createVertexClass('Fhaculty\Graph\Edge\Directed');
```

makes the builder fail or succeed.

Patching like done below helps

```
@@ -76,7 +76,13 @@ class ClassDiagramBuilder
         } else {
             $reflection = new ReflectionClass($class);
         }
-        $vertex = $this->graph->createVertex($class);
+        try {
+            $vertex = $this->graph->createVertex($class);
+        }
+        catch (Exception $exc) {
+          return;
+        }
+
```

but what is the root cause?

How can I create a patch for this issue through git? I found https://gist.github.com/piscisaureus/3342247 but we have no PRs yet.

**Comment by clue** [comment]
Copying my lost comment from the old ticket:

> Thanks for your report.
> 
> I'm not entirely sure if that's the right way to do it either.
> 
> $loader->createVertexClass('A');
> $loader->createVertexClass('A');
> 
> In accordance to how Graph::createVertex() works, you'd probably expect it to fail because you're using a duplicate ID, right? That's the current behavior, so far.
> 
> But the issue probably stems from the fact that createVertexClass() also automatically creates vertices for super classes. This may not be entirely obvious, but makes the API more convenient to use.
> 
> So your above code can be fixed by adding a hasVertexClass() check before calling createVertexClass(). IMHO this sounds sane, but it's a bit inconvenient.
> 
> So maybe we should provide an additional convenience method instead? Perhaps something like createVerticesClasses(array $classes) that does all this magic? Also, how to proceed with additional super classes not given in the array?

TLDR: 

The problem stems from the fact that `createVertexClass()` also automatically adds (missing) parent classes.

**Comment by clue** [comment]
Just _ignoring_ the call sounds like a bad approach, IMHO. Perhaps we can some up with a cleaner design instead?

We haven't approached v0.1 yet, so I'm perfectly fine with breaking the existing API.

> How can I create a patch for this issue through git?

Fork, patch, commit, push and file a new PR and make sure to include something along the lines of "Fixes #2" in the description :)

**Comment by clemens-tolboom** [comment]
I'm fine with the parents added automatically. That's an awesome feature.

I assumed I was a contrib for this module ... just tried to push and failed ... so now cloning

**Comment by clue** [comment]
> I assumed I was a contrib for this module

So did I :) Just added your contributor flag.

Handling _most_ of this through PRs/tickets still sounds like a good idea though :)

**Comment by clemens-tolboom** [comment]
In Drupal if have 2 classes and when rendering them in the same diagram I i get the following

```
[classdiagram
\Drupal\Core\Plugin\ContextAwarePluginBase
\Drupal\Component\Plugin\ContextAwarePluginBase
]
```

give

```
ID must be unique
```

tobecontinued

**Comment by clue** [comment]
I assume this boils down to this (pseudo code):

``` php
class A { }
class B extends A { }

$g->addClassAndItsParents('B'); // will add B and its parent A
$g->addClassAndItsParents('A'); // fails, as A is already added
```

Could you confirm this? If so, I think this makes sense from an API point of view.

However, I absolutely also see the need for a method that has the semantic of a `makeSureClassAndItsParentsAreAdded('A')` kind of call which only adds a class if it's not already added.

A PR would be very much appreciated :) Otherwise, I can look into this, but expect this to take some time.

**Comment by clemens-tolboom** [comment]
I'm loosing track of what's in and what's not ... indeed the Component extends the Core.

Don't spent time on this issue now. I as a user of graph-uml really appreciate tagged releases first :-p

/me where are those issues ~:-(

**Comment by clemens-tolboom** [comment]
I'm confused.

From a user point of view I just have a dirty list of classes and I want to see the diagram.

I asked for adding this with ID `B` with options `add-parents`

Next adding `A` gives `ID must be unique`

For a UML diagram generated through reflection adding an existing ID should be ok.

So for class ClassDiagramBuilder the code change still hold.

``` php
-        $vertex = $this->graph->createVertex($class);
+        $vertex = $this->graph->createVertex($class, TRUE);
```

For a utility tool editing classes it would be invalid.

**Comment by clue** [comment]
I think your issue has been resolved in PR #26 and this ticket can safely be closed. Otherwise please comment so we can re-open this.
